module/zfs/dsl_bookmark.c

```diff
@@ -413,6 +456,219 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
 	    fnvlist_num_pairs(bmarks), ZFS_SPACE_CHECK_NORMAL));
 }
 
+static int
+dsl_bookmark_clone_check_impl(nvpair_t *pair, dmu_tx_t *tx)
+{
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	int error;
+	zfs_bookmark_phys_t bmark_phys = { 0 };
+
+	/* Verify new and source bookmark strings pass bookmark namecheck */
+	{
+		char *source;
+		if (bookmark_namecheck(nvpair_name(pair), NULL, NULL) != 0)
+			return (SET_ERROR(EINVAL));
+		source = fnvpair_value_string(pair);
+		if (source == NULL)
+			return (SET_ERROR(EINVAL));
+		if (bookmark_namecheck(source, NULL, NULL) != 0)
+			return (SET_ERROR(EINVAL));
+	}
+
+	/* Verify that new and source bookmark are on the same dataset */
```

Comment:
They shouldn't have to be on the same dataset.  Like normal bookmark creation, the source can be any point in the history of the filesystem that will contain the new bookmark.  i.e. it can be from the origin.  This can be checked with `dsl_dataset_is_before()`, like we do for normal bookmark creation.  Or better yet, by using` dsl_bookmark_lookup()` which will call `dsl_dataset_is_before()` for us.

If this is not the case, the ioctl interface should be changed so that userland isn't required to say the exact same thing (the name of the filesystem) twice.  i.e. the value would omit the filesystem name.

Reply:
> If this is not the case, the ioctl interface should be changed so that userland isn't required to say the exact same thing (the name of the filesystem) twice. i.e. the value would omit the filesystem name.

I think it is safer to leave as much string expansion logic as possible in userspace or a safe language like Lua, and have the C interface always expect full paths, albeit allowing the 'different source dataset' scenario you outlined above.

To be clear, given the plan to merge `dsl_bookmark_clone_*` and `dsl_bookmark_create_*` machinery, the nvlist that both zcp and ioctl will pass to `dsl_bookmark_create` will look like this (agreed?):

```
{
        // with source on origin
        "tank/fs1#bm1_clone" : "tank/fs1_origin#bm1",
        "tank/fs1#newbm_of_a" : "tank/fs1_origin@a",

        // with source on same dataset: still have full paths (expansion done by caller, see zfs_do_bookmark)
        "tank/fs1#b" : "tank/fs1@b",
        "tank/fs1#aboomkark_clone" : "tank/fs1#abookmark",
}
```

Reply:
FYI: the proposal above is what's implemented in https://github.com/zfsonlinux/zfs/commit/fa99694ea55fb6a1477e2804a7e150bb31e3b84a see function `dsl_bookmark_create_nvl_validate`